src/google/protobuf/compiler/csharp/csharp_enum_field.cc

```diff
@@ -184,13 +176,100 @@ void EnumFieldGenerator::WriteEquals(Writer* writer) {
   }
 }
 void EnumFieldGenerator::WriteToString(Writer* writer) {
+  writer->WriteLine("PrintField(\"$0$\", $1$, $2$_, writer);",
+                    descriptor_->name(), has_property_name, name());
+}
+
+EnumOneofFieldGenerator::EnumOneofFieldGenerator(const FieldDescriptor* descriptor,
+						 int fieldOrdinal)
+  : EnumFieldGenerator(descriptor, fieldOrdinal) {
+  oneof_name =
+    UnderscoresToCamelCase(descriptor->containing_oneof()->name(), false);
+  oneof_property_name =
+    UnderscoresToCamelCase(descriptor->containing_oneof()->name(), true);
+
+  has_property_name = oneof_name + "Case_ == " + oneof_property_name +
+    "OneofCase." + property_name();
+  
+}
+
+EnumOneofFieldGenerator::~EnumOneofFieldGenerator() {
+}
+
+void EnumOneofFieldGenerator::GenerateMembers(Writer* writer) {
+  AddDeprecatedFlag(writer);
   if (SupportFieldPresence(descriptor_->file())) {
-    writer->WriteLine("PrintField(\"$0$\", has$1$, $2$_, writer);",
-                      descriptor_->name(), property_name(), name());
-  } else {
-    writer->WriteLine("PrintField(\"$0$\", $1$_, writer);",
-                      descriptor_->name(), name());
+    writer->WriteLine("public bool Has$0$ {", property_name());
+    writer->WriteLine("  get { return $0$; }", has_property_name);
+    writer->WriteLine("}");
+  }
+  AddPublicMemberAttributes(writer);
+  writer->WriteLine("public $0$ $1$ {", type_name(), property_name());
+  writer->WriteLine("  get { return $0$ ? ($1$) $2$_ : $3$; }",
```

Comment:
Fair enough, I am fine with leaving as it for now.
We should definitely consider doing some cleanup in the future.

Regarding changing writer to io::Printer  - I originally tried to do that as well, but it was so much work that that I eventually gave up. Now when we have more tests coverage, it might be easier. Another problem was that the readability wasn't great - because with io:Printer you have to name your variables, which brings even more boilerplate.

I would suggest migrating to io::Printer once we've done some slimming of the generated code, so that there's less work to do.